### PR TITLE
Implement blocking RawTextReader wrapper

### PR DESCRIPTION
**Pull request by nirosys**
*Issue #, if available:* n/a

*Description of changes:*
This PR adapts the blocking `RawTextReader` to act as a blocking wrapper around the non-blocking `RawTextReader`.

The blocking reader uses a non-blocking reader with a `Vec<u8>` backed text buffer, starting with a default size of 4kb. Data is read from the IonDataSource into the text buffer maintaining the 4kb size unless more data is needed in order to parse a larger value, such as a blob.

With each data consuming call, the reader will perform a check if the underlying source has more data, and if the current text buffer has less than 25% data remaining. If so, then the reader will fill the text buffer before calling the consuming function on the non-blocking reader.

If an incomplete error is returned during a consuming call, the blocking reader will attempt to fill more data into the text buffer. The reader will continue to feed data into the text buffer, until there is no more data, or the consuming function no longer returns an incomplete error. The incomplete error is bubbled up to the caller if the source data is exhausted. 

#### Potential Addition
Right now the blocking reader will use a 4kb buffer unless more space is needed to parse larger values. This is something the user should have control over, and probably warrants having a `with_capacity`-style creation method. I can follow up with a PR to add that too.

#### Discovered Issue
While extending the unit tests in order to pass the reported codecov %s, I ran into an issue where parsing a struct with 4k spaces between the last value and the closing struct delimiter failed. Adding a comma to end the last value resulted in a correct parse. Removing the spaces, also resulted in a correct parse. I updated the code to force newly read characters to be added to the current loaded lines (to see if it was a buffer reading issue), which ensured the closing delimiter was in the buffer when parsing and even saw the delimiter in the error message, but the failure to parse continued.

I'll dig more into this to figure out what the underlying issue is. Adding the `with_capacity`-style method above will help to determine if this is due to this PR's buffer reading, or some other issue. 

#### Update
The most recent commit addresses feedback on this PR, as well as a couple bugs that were discovered while implementing the end of stream signal.

Part of the fixes include adding a "carried state", which is a mechanism to allow tracking of state for more complex `next` logic. The first place this is used is parsing of struct fields where the field name is parsed, then the value, which can fail when an incomplete text error is reached while parsing the value.

The second place is when stepping out of a container. After popping the container from the parent list, if the reader is within another container it needs to find the delimiter or container end. During this, an incomplete text error can be reached, so state needs to be carried for the next call to `next`, after the reader gets more data.

In addition to the bug fixes, the stream end marker was added so that the reader can rely on Incomplete Text errors, rather than trying to read more data when the buffer starts getting low.

#### Update 2 
This round of updates addresses the issues brought up in review. Most of note is the handling of errors in the non-blocking `step_out` method. Details about the fix are left as comments within the `step_out` function.

Another note for this update is that I've included line, and column, numbers for errors that previously had constant placeholders. The implementation for column numbers is currently not accurate and due to recent changes in restack, has the potential to be altered when restacking, causing non-zero column numbers to be shifted by previously consumed lengths. I plan on following up soon with an update to handle lines and columns correctly.

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by nirosys** [comment]
Moved this out of draft and into review. I cannot reproduce the coverage %s that codecov is generating. Running the github actions locally gives a 99.06% coverage for the blocking `raw_text_reader.rs` (which is most of this PR). I'm not entirely sure how codecov.io is producing the patch %, but the lines it is marking in the PR source as not covered include comments and blank lines.

File: src/text/raw_text_reader.rs
Comment (zslayton) [inline src/text/raw_text_reader.rs:102]: Rather than using the (reasonable!) heuristic you've implemented on each call to `next()`, I propose:
1. Doing an initial read at reader creation time. Our [builder APIs return an `IonResult<ReaderType>`](https://github.com/amzn/ion-rust/blob/94c83bf336f0c3902ac81b594f5d38774c097782/src/reader.rs#L37) to accommodate I/O during construction. (In that particular example, we peek at the first four bytes of the file to detect whether we're dealing with text or binary.)
2. Only reading from the source when we encounter an `Incomplete`.

That way most calls to `next()` will be free in terms of I/O checking.

File: src/text/raw_text_reader.rs
Comment (zslayton) [inline src/text/raw_text_reader.rs:110]: Why 1024 instead of whatever the buffer has room for?

File: src/text/raw_text_reader.rs
Comment (zslayton) [inline src/text/raw_text_reader.rs:104]: Here's a slightly more concise version if you like:
```rust
fn next(&mut self) -> IonResult<RawStreamItem> {
    self.feed_if_needed()?;
    loop {
        let mut result = self.reader.next();
        if let IonError::IncompleteText { .. } = result {
            if 0 == self.read_source(1024)? {
                return Err(err);
            }
       } else {
            return result;
        }
}
```

File: src/text/raw_text_reader.rs
Comment (zslayton) [inline src/text/raw_text_reader.rs:237]: I suggest making a `retry_on_incomplete` helper method that takes a closure to loop over as long as it's returning `Incomplete`. Then we can use it in `step_in()`, `step_out()` and (when this wrapper is format-agnostic) all of the `read_*` calls that can trigger I/O in the binary reader.

**Review by zslayton — COMMENTED** [review]
I love a good mostly-deletions PR! 🎉 

image

This blocking wrapper is _tantalizingly close_ to being able to work for any format of raw reader, which is where I'd like to get to. In a follow-on PR, could you make it generic over another `R: RawReader`? Then we can delete the blocking raw binary reader implementation too! (There are probably prerequisite changes that need to happen for this; in particular, I think we'll need to consolidate `Incomplete` and `IncompleteText` into a single error type. This in turn means putting some more thought into how we report reader position in a format-agnostic way.)

Some minor comments below.

File: src/text/non_blocking/text_buffer.rs
Comment (nirosys) [reply src/text/non_blocking/text_buffer.rs:287]: oof.. Thank you for calling that out. I had run into an issue where parsing was failing due to unexpected characters in the text (large number of '\0's), that was ultimately due to the `read_from` reserving the needed space for the requested bytes. I quickly fixed it with this resize, to truncate the buffer down to only what was read.  At the time I was focused on another area of the code, and never went back to investigate why the issue was happening.

I've dug deeper now, and I'll post a new commit that removes this resize in favor of fixing the underlying bug in the text buffer. 

When UTF-8 validation is done, my original code validated `&self.data.as_ref()[valid_end..]`. Once a `read_from` happens that pre-extends the buffer beyond the actual data size the UTF-8 validation starts validating the end of the buffer without read data. Since the data is initialized with '\0's, the unread bytes do not get evaluated by `from_utf8`, however the `Ok(_)` match assumed the entire buffer was valid and would return the buffer's length as the validated end.

File: src/text/raw_text_reader.rs
Comment (nirosys) [reply src/text/raw_text_reader.rs:102]: I think we need a way for the user to flag the non-blocking reader as 'done' in terms of data being provided. The reason I started reading ahead of consuming functions was to reduce the chances of the parser returning back values and advancing when partial values were present.

For example, in the unit tests that used the builder APIs, the chain of first 4 bytes followed by the rest of the buffer, resulted in a first `read` of 4 bytes. For text encoded timestamps, that lined up perfectly with the 4 digit year, which would parse as an integer. Then the next read would generate an error. Those boundaries seem to be possible anywhere, and there seems to be ample opportunities for partial values to be parsed as valid values.

What do you think of having the user signal when data is complete, and if that signal has not been made and a parse would consume the remainder of the buffer, an incomplete error is returned and the data is left as is?

With that I think we can rely on the `Incomplete` to signal when to read.

Comment (zslayton) [reply src/text/raw_text_reader.rs:102]: Oh! Thank you for explaining the problem. Take a look at the old `RawTextReader`'s [`parse_value_at_eof()` method](https://github.com/amzn/ion-rust/blob/main/src/text/raw_text_reader.rs#L376). What we need (and I think this is similar to what you described) is a way for the blocking reader to say:
```
if self.reader.next() == Incomplete and self.input().is_eof() {
  return self.reader.parse_value_at_eof();
}
```
What I don't know (and we don't necessarily need to solve this here/now) is how to do this in a way that's format agnostic. The binary reader doesn't have this problem; it can always tell where the boundaries of a value are.

We may end up needing to add some methods to the `RawReader` trait.

File: src/text/raw_text_reader.rs
Comment (nirosys) [reply src/text/raw_text_reader.rs:110]: Arbitrarily chose 1024, but it isn't tied to the remaining space in the buffer in order to allow the buffer to grow to support reading blobs that are larger than the buffer.

Comment (zslayton) [reply src/text/raw_text_reader.rs:110]: I wonder if we should double it on each retry so it ramps up faster for large values?

File: src/text/raw_text_reader.rs
Comment (nirosys) [reply src/text/raw_text_reader.rs]: With the recent commit, the retry logic is only contained in the `next`, and `step_out` methods. I do want to pull it into a helper method, but ran into issues due to borrows of self. I need to spend a little more time with it, but didn't want to hold up the PR any more than it already has been.

File: src/text/non_blocking/raw_text_reader.rs
Comment (zslayton) [inline src/text/non_blocking/raw_text_reader.rs:801]: A bit simpler:
```suggestion
        if self.carried_state != Some(ReaderState::ContainerDelimiter) {
```

File: src/text/non_blocking/raw_text_reader.rs
Comment (zslayton) [inline src/text/non_blocking/raw_text_reader.rs:828]: This strikes me as late in the method to set `carried_state`; what if the loop on line 815 encounters incomplete data while it's exhausting all of the child elements of the container we're stepping out of?

File: src/text/raw_text_reader.rs
Comment (zslayton) [inline src/text/raw_text_reader.rs:45]: Is this intentionally `pub`? I expected the blocking implementation to handle this automatically.

File: src/text/raw_text_reader.rs
Comment (zslayton) [inline src/text/raw_text_reader.rs:93]: We're willing to read by 10x the `max_read_size`, so I think we should reconsider that variable name. `original_buffer_size`, maybe?

File: src/text/raw_text_reader.rs
Comment (zslayton) [inline src/text/raw_text_reader.rs:769]: Could `TestIonSource` be replaced by [`io::Cursor`](https://doc.rust-lang.org/std/io/struct.Cursor.html) from the stdlib?

File: src/text/raw_text_reader.rs
Comment (zslayton) [inline src/text/raw_text_reader.rs:856]: Is this `repro` test a debugging leftover?

File: src/text/non_blocking/raw_text_reader.rs
Comment (zslayton) [inline src/text/non_blocking/raw_text_reader.rs:34]:
```suggestion
    // A previous `next` call popped the container, but was unable to find an outer container's delimiter or end.
    ContainerDelimiter,
```

File: src/text/non_blocking/raw_text_reader.rs
Comment (zslayton) [inline src/text/non_blocking/raw_text_reader.rs:32]: My initial thought here was that we should just make parsing a field name and value transactional; if either fails we just bubble up `Incomplete` and retry both on the next attempt.

We still need to track state for stepping out of a container, so we could keep this implementation even though it's a bit more complex. Interested to hear your thoughts, though.

File: src/text/non_blocking/raw_text_reader.rs
Comment (zslayton) [inline src/text/non_blocking/raw_text_reader.rs:55]: I think we can merge a couple of these member fields; rather than holding a `carried_state: Option<ReaderState>`, I think it can have a `state: ReaderState` with a new variant that represents "ready to read the next value".

See the non-blocking raw binary reader's [`ReaderState` type](https://github.dev/amzn/ion-rust/blob/b01a603216edfe83d51aa57647ea912665f5334b/src/binary/non_blocking/raw_binary_reader.rs#L233-L261) for comparison.

File: src/text/non_blocking/raw_text_reader.rs
Comment (zslayton) [inline src/text/non_blocking/raw_text_reader.rs:50]: I realize this isn't part of your PR, but I wonder if this `is_eof` needs to be tracked separately from just checking `self.buffer.is_empty()` anywhere we currently read it, perhaps via a helper method with a friendly name.

File: src/raw_reader.rs
Comment (zslayton) [inline src/raw_reader.rs:12]: I think I prefer the phrasing in your comment. How about `stream_complete`?
```suggestion
    fn stream_complete(&mut self);
```

File: src/text/non_blocking/raw_text_reader.rs
Comment (zslayton) [reply src/text/non_blocking/raw_text_reader.rs:32]: > What is the benefit of throwing the field name away? Is it just to keep the code simple?

Yep!

> Even if it is only during read boundaries, the data, and read/server response/etc size play into how impactful that is.

Intuitively, I'd expect the work wasted at read boundaries to be negligible in practice. I'd prefer to start simple and add additional state machinery if/when it shows up in profiling.

File: src/text/non_blocking/raw_text_reader.rs
Comment (zslayton) [inline src/text/non_blocking/raw_text_reader.rs:575]: Is the `(0, 0)` a placeholder?

File: src/text/non_blocking/raw_text_reader.rs
Comment (zslayton) [inline src/text/non_blocking/raw_text_reader.rs:885]: I don't think I understand this bit -- let's talk through it offline.

File: src/text/non_blocking/raw_text_reader.rs
Comment (zslayton) [inline src/text/non_blocking/raw_text_reader.rs:1419]: Could you add a comment showing where 47 is in the input? I see it's somewhere in the string, but knowing exactly where would help me mentally run the unit test.

File: src/text/non_blocking/raw_text_reader.rs
Comment (zslayton) [inline src/text/non_blocking/raw_text_reader.rs:1438]: If I'm at depth 3 and `step_out` fails, then the next time I call `step_out` successfully I should be at depth 2. If I'm understanding this correctly, the current implementation allows the reader to make _some_ progress before failing and the user can then resume using it from that unknown interstitial position. I think there's a lot of room for confusion here (how far did the reader get? where is it now?), and it's not consistent with our other `Reader` implementations' contracts.

File: src/text/non_blocking/text_buffer.rs
Comment (zslayton) [inline src/text/non_blocking/text_buffer.rs:312]: If you prefer:
```suggestion
        self.data_exhausted = self.data_end > 0;
```

File: src/text/raw_text_reader.rs
Comment (zslayton) [inline src/text/raw_text_reader.rs:30]: Can you open an issue to track exposing this to the `Builder` API?

File: src/text/raw_text_reader.rs
Comment (zslayton) [inline src/text/raw_text_reader.rs:52]: I can't remember, is this information not already tracked in the raw reader or text buffer?

File: src/text/raw_text_reader.rs
Comment (zslayton) [reply src/text/raw_text_reader.rs:762]: You can create a reader that uses a `&mut Cursor` and then get the cursor back when you're done. Personally, I'd also be in favor of having a `Reader::input_source()` or something that lent out a reference to the underlying data source. We don't have to solve that in this PR, though.

File: src/text/non_blocking/raw_text_reader.rs
Comment (nirosys) [reply src/text/non_blocking/raw_text_reader.rs:267]: The `ContainerDelimiter` state is specifically for when we have finished reading all of the container's items and have stepped out of the container, but need to find an inner delimiter, or closing delimiter, for a parent container. That is why it is set after exhausting the contents, and after we test for an empty `parents` list.

If an error occurs while reading the container's values, like your example, the carried state *needs* to be the state from where the error occurred. Since the calls to `next` in the exhaustion loop can step into other containers, or fail to parse any value, we can't set a carried state until we're done exhausting the contents and (from the perspective of the step_out) are back in control of what's being parsed.

In the case of `ContainerDelimiter`, if we set it before reading, and an error occurs while reading a value, the next step into `step_out` would cause the reader to start looking for a parent container's delimiter. If we set it to any other state, then another instance of `step_out` could clobber that state with whatever it was doing at the time.

I've added a unit test for your example and will add some more, and push them up when they're done.

File: src/text/non_blocking/raw_text_reader.rs
Comment (nirosys) [reply src/text/non_blocking/raw_text_reader.rs:885]: 👍

File: src/text/non_blocking/raw_text_reader.rs
Comment (nirosys) [reply src/text/non_blocking/raw_text_reader.rs:237]: You're right. I had originally set it to the depth step_out was called at and did not update the comment when I changed to the target depth.

File: src/text/non_blocking/raw_text_reader.rs
Comment (nirosys) [reply src/text/non_blocking/raw_text_reader.rs:1438]: Sorry for the sporadic responses, I let this one marinate for a bit. Your comment about the reader implementations' contracts is very surprising.

Confusion for whom? The user? Or maintainers?
What reader contract? I may have missed documentation describing this.

If the intent truly results in requiring the entire remaining content of a container to be in memory before a `step_out` can actually make progress, and re-parsing the same data until it does, then I have concerns. Let's discuss.

Comment (nirosys) [reply src/text/non_blocking/raw_text_reader.rs:1438]: Update after our discussion:

I misunderstood Zach's comment. We do not need to roll back to where the reader was when `step_out`/`next` was called, we need to ensure that if we cannot complete the `step_out` or `next` then we do not allow the user to advance the reader in any other way until the initial failed call is completed. This should match the behavior of the binary reader.

File: src/text/non_blocking/raw_text_reader.rs
Comment (nirosys) [reply src/text/non_blocking/raw_text_reader.rs:196]: I will get this implemented with the next update.

File: src/text/raw_text_reader.rs
Comment (nirosys) [reply src/text/raw_text_reader.rs:52]: Yea, this was missed in the last update. I had removed the original use of `bytes_fed`, and the function that exposed it, but missed the actual field definition and update to it. Deleted.

File: src/text/raw_text_reader.rs
Comment (nirosys) [reply src/text/raw_text_reader.rs:30]: Absolutely

File: src/text/non_blocking/raw_text_reader.rs
Comment (zslayton) [inline src/text/non_blocking/raw_text_reader.rs:38]: Hm, I'm unsure what "looking for the parent container" means here. It sounds distinct from expecting the closing delimiter. Maybe you could add an example?

File: src/text/non_blocking/raw_text_reader.rs
Comment (zslayton) [reply src/text/non_blocking/raw_text_reader.rs:75]: I don't think I've grasped why this couldn't be stored in the `SteppingOut` variant, but we don't need to hold up the PR for it.

**Review by zslayton — APPROVED** [review]

**Pull request merged**
